File: osmflat/examples/cpp/pub_names.cpp
#include "osm_generated.h"
#include <cstring>
#include <iostream>

bool comp(const char *left, const char *right) {
  while (*right == *left) {
    if (*right == 0) {
      return true;
    }
    left++;
    right++;
  }
  return false;
}

void print_pubs(const osm::Osm &archive, std::pair<uint64_t, uint64_t> range) {
  auto tags = archive.tags();
  auto tags_index = archive.tags_index();
  const char *strings = archive.stringtable().char_ptr();

  auto is_pub = [&](std::pair<uint64_t, uint64_t> range) {
    for (auto idx : tags_index.slice(range)) {
      auto tag = tags[idx.value];
      const char *key = strings + tag.key_idx;
      if (comp(key, "amenity")) {
        const char *value = strings + tag.value_idx;
        return comp(value, "pub");
      }
    }
    return false;
  };

  if (!is_pub(range)) {
    return;
  }

  auto get_name = [&](std::pair<uint64_t, uint64_t> range) {
    for (auto idx : tags_index.slice(range)) {
      auto tag = tags[idx.value];
      const char *key = strings + tag.key_idx;
      if (comp(key, "name")) {
        return strings + tag.value_idx;
      }
    }
    return "unknown pub name";
  };

  std::cout << get_name(range) << std::endl;

  for (auto idx : tags_index.slice(range)) {
    auto tag = tags[idx.value];
    const char *key = strings + tag.key_idx;
    if (std::strncmp(strings + tag.key_idx, "addr:", 5) == 0) {
      const char *value = strings + tag.value_idx;
      std::cout << "  " << key << ": " << value << std::endl;
    }
  }
}

int main(int argc, char const *argv[]) {
  if (argc != 2) {
    std::cerr << "USAGE: pub_name <osmflat-archive>" << std::endl;
    return 1;
  }

  auto storage = flatdata::FileResourceStorage::create(argv[1]);
  auto archive = osm::Osm::open(std::move(storage));

  for (auto node : archive.nodes()) {
    print_pubs(archive, node.tags);
  }

  for (auto way : archive.ways()) {
    print_pubs(archive, way.tags);
  }

  return 0;
}

File: osmflat/examples/pub-names.rs
//! Shows the names and addresses of all pubs.
//!
//! Demonstrates
//!
//!  * iteration through tags belonging to a node and a way
//!  * accessing of tags by key
//!  * filtering of tags
//!
//! LICENSE
//!
//! The code in this example file is released into the Public Domain.

use osmflat::{find_tag, has_tag, iter_tags, FileResourceStorage, Osm};
use std::str;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let archive_dir = std::env::args()
        .nth(1)
        .ok_or("USAGE: pub_names <osmflat-archive>")?;
    let archive = Osm::open(FileResourceStorage::new(archive_dir))?;

    let nodes_tags = archive.nodes().iter().map(|node| node.tags());
    let ways_tags = archive.ways().iter().map(|way| way.tags());

    for tag_range in nodes_tags.chain(ways_tags) {
        if has_tag(&archive, tag_range.clone(), b"amenity", b"pub") {
            let name = find_tag(&archive, tag_range.clone(), b"name");
            let name = name.map(|s| str::from_utf8(s).unwrap_or("broken pub name"));
            println!("{}", name.unwrap_or("unknown pub name"));

            let addrs = iter_tags(&archive, tag_range).filter(|(k, _)| k.starts_with(b"addr:"));
            for (k, v) in addrs {
                if let (Ok(addr_type), Ok(addr)) = (str::from_utf8(k), str::from_utf8(v)) {
                    println!("  {addr_type}: {addr}");
                }
            }
        }
    }

    Ok(())
}
